### PR TITLE
[SPARK-33095] Follow up, support alter table column rename.

### PR DESCRIPTION
**Pull request by ScrapCodes**
### What changes were proposed in this pull request?

Support rename column for mysql dialect.

### Why are the changes needed?

At the moment, it does not work for mysql version 5.x. So, we should throw proper exception for that case.

### Does this PR introduce _any_ user-facing change?

Yes, `column rename` with mysql dialect should work correctly.

### How was this patch tested?

Added tests for rename column.
Ran the tests to pass with both versions of mysql.

* `export MYSQL_DOCKER_IMAGE_NAME=mysql:5.7.31`

* `export MYSQL_DOCKER_IMAGE_NAME=mysql:8.0`

### CODE REVIEW

**Comment by ScrapCodes** [comment]
While testing #30038, I found rename column needs some work for mysql dialect. More details added in the code comments.

cc @cloud-fan and @huaxingao

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala [hunk 1]
```diff
@@ -77,11 +86,26 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite with V2JDBCTest {
     assert(msg1.contains("Cannot update alt_table field ID: string cannot be cast to int"))
   }
 
+  override def testRenameColumn(tbl: String): Unit = {
+    if (db.imageName.matches(".*mysql.5\\.[0-9].*")) {
```
Comment (cloud-fan) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:90]: we can check the config value here.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala [hunk 2]
```diff
@@ -77,11 +86,26 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite with V2JDBCTest {
...
+      sql(s"CREATE TABLE $tbl (ID STRING NOT NULL) USING _")
+      // Update nullability is unsupported for mysql db.
```
Comment (cloud-fan) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:92]: rename is unsupported ...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -2858,6 +2858,14 @@ object SQLConf {
     .stringConf
     .createWithDefault("")
 
+  val JDBC_MYSQL_VERSION =
+    buildConf("spark.sql.jdbc.mysql.version")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2862]: what if users want to connect to both mysql 5 and 8? I think it's better to make it a catalog option, not a session config.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 4]
```diff
@@ -57,6 +58,25 @@ private case object MySQLDialect extends JdbcDialect {
...
+  // See Old Syntax: https://dev.mysql.com/doc/refman/5.6/en/alter-table.html
+  // According to https://dev.mysql.com/worklog/task/?id=10761 old syntax works for
+  // both versions of MySQL i.e. 5.x and 8.0
+  // The old syntax requires us to have type definition. Since we do not have type
+  // information, we throw the exception.
+  override def getRenameColumnQuery(
+      tableName: String,
+      columnName: String,
+      newName: String): String = {
+    if (SQLConf.get.jdbcMySQLVersion.matches("^8\\.[0-9].*")) {
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:70]: Instead of using a session config, we probably want to get database version using `DatabaseMetaData` in `JdbcUtils.altertable` and pass the info here, and only do `RENAME` if version >= 8
```
conn.getMetaData.getDatabaseMajorVersion
```

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:70]: ah, it's even better if we can get the database version, so that users don't need to specify it.

Comment (ScrapCodes) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala]: Thank you @huaxingao and @cloud-fan, for this suggestion.

**Comment by ScrapCodes** [comment]
@cloud-fan and @huaxingao I have updated the PR with your suggestions and also rebased it with master. Please take a look again.

**Comment by ScrapCodes** [comment]
Jenkins, retest this please

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala [hunk 5]
```diff
@@ -50,16 +50,23 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite with V2JDBCTest {
...
     override def getJdbcUrl(ip: String, port: Int): String =
-      s"jdbc:mysql://$ip:$port/mysql?user=root&password=rootpass"
+      s"jdbc:mysql://$ip:$port/" +
+        s"mysql?user=root&password=rootpass&allowPublicKeyRetrieval=true&useSSL=false"
   }
 
-  override def sparkConf: SparkConf = super.sparkConf
-    .set("spark.sql.catalog.mysql", classOf[JDBCTableCatalog].getName)
-    .set("spark.sql.catalog.mysql.url", db.getJdbcUrl(dockerIp, externalPort))
+  override def sparkConf: SparkConf = {
```
Comment (cloud-fan) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:57]: unnecessary change

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala [hunk 6]
```diff
@@ -77,11 +84,25 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite with V2JDBCTest {
     assert(msg1.contains("Cannot update alt_table field ID: string cannot be cast to int"))
   }
 
+  override def testRenameColumn(tbl: String): Unit = {
+    assert( mySQLVersion > 0)
```
Comment (cloud-fan) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MySQLIntegrationSuite.scala:88]: nit: `assert(mySQLVersion > 0)` (remove extra space)

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 7]
```diff
@@ -46,6 +46,20 @@ trait V2JDBCTest extends SharedSparkSession {
...
   }
 
+  def testRenameColumn(tbl: String): Unit = {
+    sql(s"CREATE TABLE $tbl (ID STRING NOT NULL, ID1 STRING NOT NULL) USING _")
+    sql(s"ALTER TABLE $tbl RENAME COLUMN ID TO ID2")
+    val t = spark.table(s"$tbl")
+    val expectedSchema = new StructType().add("ID2", StringType, nullable = true)
+      .add("ID1", StringType, nullable = true)
+    assert(t.schema === expectedSchema)
+    // Rename to already existing column
```
Comment (cloud-fan) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:56]: This can be put in the `test("SPARK-33034: ALTER TABLE ... rename column")`, as the behavior should always be the same.

**Review by cloud-fan — APPROVED** [review]

**Comment by huaxingao** [comment]
@ScrapCodes seems you forgot to address this https://github.com/apache/spark/pull/30142#discussion_r514212732. Everything else is good.

**Review by huaxingao — APPROVED** [review]

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala (hunk 7 above)
Comment (ScrapCodes) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala]: done

**Comment by ScrapCodes** [comment]
Hi @huaxingao and @cloud-fan, Please take a look.

**Comment by cloud-fan** [comment]
thanks, merging to master!
